docs/rfcs/0003-initial.md

```diff
@@ -0,0 +1,280 @@
+  > with a very different feel, and often a non-Go-like feel.
+> - Feel like Go.
+> - Care mostly about the common cases. Common SQL should be portable. SQL edge
+  > cases or db-specific extensions can be detected and conditionally used by
+  > the application. It is a non-goal to care about every particular db's
+  > extension or quirk.
+> - Separate out the basic implementation of a database driver (implementing the
+  > `sql/driver` interfaces) vs the implementation of all the user-level types
+  > and convenience methods. In a nutshell:<br> User Code &rarr; `sql` package
+  > (concrete types) &rarr; `sql/driver` (interfaces)<br> Database Driver &rarr;
+  > `sql` (to register) + `sql/driver` (implement interfaces)
+> - Make type casting/conversions consistent between all drivers. To achieve
+  > this, most of the conversions are done in the `sql` package, not in each
+  > driver. The drivers then only have to deal with a smaller set of types.
+> - Be flexible with type conversions, but be paranoid about silent truncation
+  > or other loss of precision.
+> - Handle concurrency well. Users shouldn't need to care about the database's
+  > per-connection thread safety issues (or lack thereof), and shouldn't have to
+  > maintain their own free pools of connections. The `sql` package should deal
+  > with that bookkeeping as needed. Given an `*sql.DB`, it should be possible
+  > to share that instance between multiple goroutines, without any extra
+  > synchronization.
+> - Push complexity, where necessary, down into the `sql`+`driver` packages,
+  > rather than exposing it to users. Said otherwise, the `sql` package should
+  > expose an ideal database that's not finnicky about how it's accessed, even
+  > if that's not true.
+> - Provide optional interfaces in `sql/driver` for drivers to implement for
+  > special cases or fastpaths. But the only party that knows about those is the
+  > sql package. To user code, some stuff just might start working or start
+  > working slightly faster.
+
+## Goals
+
+For this initial v0.1 release, we would like to provide the minimal set of core
+capabilities that are enough to be useful, even if they may not be as
+comprehensive, flexible, efficient, or convenient as we'd like.
+
+Prior to v1.0, we do not make any guarantees about API stability
+([as per semver](https://semver.org/spec/v2.0.0.html#spec-item-4)), but we
+should still be thoughtful and try to avoid unnecessary breakage.
+
+## Database Driver API (`x/database/sql/driver.ts`)
+
+Inspired by [`driver.go`](https://golang.org/src/database/sql/driver/driver.go),
+`./driver.ts` exports interfaces for database driver libraries to implement for
+interoperability with `./sql.ts`'s user interface. **Most users should have no
+reason to use these interfaces directly.** Note that drivers do not "register"
+themselves with this library as they do in with the Go package.
+
+Many of these interfaces define methods in both optional async and sync
```

Comment:
Should we clarify expectations around this? For example: do the implementations need to be consistent?

E.g.: For SQLite, you can only have one transaction per connection at a time, so a sync transaction interface might throw, while this can be nicely handled with an async transaction interface. But for the other methods, async makes no sense for SQLite. In this case, should the driver implement the async interface for transactions and the sync interface for the other methods?

(And do we want to provide a default async implementation which just warps around sync methods, if they are available, or should the consumer of the driver figure out which methods are set (and how))

Also, what are our expectations around allowing the driver implementations to throw errors, i.e. can every method trow any error it wants, or should we put restrictions on that?

Reply:
I think it would be nice to have more-specific `driver.SyncDriver` and `driver.AsyncDriver` interfaces available, so that drivers/users can be more explicit about what they support/need. I originally had them, but it was a hassle to keep updating them as the interfaces changed in other ways, so I've punted on that until the rest is more stable.

But that can probably be a type-only change. In terms of runtime implementation, `sql.Database` should raise a runtime error if the user calls a sync method and the driver's implementation is async. If they call an async method and the driver's implementation is sync, `sql.Database` will automatically fall back the sync version. Drivers shouldn't need to worry about the fallback logic themselves. There current (incomplete!) implementation has a few spots that look like this:

```ts
  // use the async method if it's defined
  const connector = await driver.openConnector?.(path) ??
    // otherwise use the sync version
    driver.openConnectorSync?.(path) ?? 
    // if neither are defined, throw an error
    notImplemented("driver missing .open[Sync] implementation");
```

So I am thinking we will allow implementations with mixed `sync` and `async` implementations for now, but we will allow users to express more specific requirements through type constraints later.

Regarding errors: good point. I haven't given it enough thought, for now I've been imagining anything can throw anything. Some structure around them would be good.

Reply:
Added a list of some possible errors and descriptions of where each may be thrown.